integration_tests/src/fixture_generator/generator.rs

```diff
@@ -0,0 +1,181 @@
+use serde::{Deserialize, Serialize};
+use winter_crypto::hashers::Blake3_256;
+use winter_math::fields::f128::BaseElement;
+
+use crate::fixture_generator::parser::Args;
+use crate::fixture_generator::writer::yaml::YamlWriter;
+use crate::fixture_generator::writer::Writer;
+
+type Blake3 = Blake3_256<BaseElement>;
+
+/// Directory state comprises all database records at a particular epoch.
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct State {
+    pub epoch: u32,
+    pub records: Vec<DbRecord>,
+}
+
+/// Delta comprises all updates published to the directory to advance to an
+/// epoch.
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct Delta {
+    pub epoch: u32,
+    pub updates: Vec<(AkdLabel, AkdValue)>,
+}
+
+/// Metadata about the output, including arguments passed to this tool and
+/// the tool version.
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct Metadata {
+    pub args: Args,
+    pub version: String,
+}
+
+// "@" has to be separated from "generated" or linters might ignore this file
+const HEADER_COMMENT: &str = concat!(
+    "@",
+    "generated This file was automatically generated by \n\
+    the fixture generator tool with the following command:\n\n\
+    cargo run -- \\"
+);
+const METADATA_COMMENT: &str = "Metadata";
+const STATE_COMMENT: &str = "State - Epoch";
+const DELTA_COMMENT: &str = "Delta - Epoch";
+
+pub async fn run() {
+    let args = Args::parse();
+    let mut rng = OsRng;
+
+    // args assertions
+    assert!(args.max_updates >= args.min_updates);
+    assert!(args
+        .capture_states
+        .as_ref()
+        .map_or(true, |states| states.iter().max().unwrap() <= &args.epochs));
+    assert!(args
+        .capture_deltas
+        .as_ref()
+        .map_or(true, |deltas| deltas.iter().max().unwrap() <= &args.epochs));
+
+    // process users
+    let mut user_map = HashMap::new();
+    for user in &args.users {
+        let mut events_map = HashMap::new();
+        for event in &user.events {
+            events_map.insert(event.epoch, event.value.clone());
+        }
+        user_map.insert(user.label.clone(), events_map);
+    }
+
+    // initialize writer
+    let buffer: Box<dyn Write> = if let Some(ref file_name) = args.out {
+        Box::new(File::create(file_name).unwrap())
+    } else {
+        Box::new(std::io::stdout())
+    };
+    let mut writer = YamlWriter::new(buffer);
+
+    // write raw args as comment
+    let raw_args = format!(
+        " {}",
+        env::args().skip(1).collect::<Vec<String>>().join(" ")
+    );
+    writer.write_comment(HEADER_COMMENT);
+    raw_args
+        .split(" -")
+        .skip(1)
+        .map(|arg| format!("  -{} \\", arg))
+        .for_each(|comment| writer.write_comment(&comment));
+
+    // write fixture metadata
+    let comment = METADATA_COMMENT.to_string();
+    let metadata = Metadata {
+        args: args.clone(),
+        version: env!("CARGO_PKG_VERSION").to_string(),
```

Comment:
So this is great for compat, but this also means we should start versioning this crate. And every commit on this crate should bump a version #. When we support reading back the metadata's we'll need versioned readers as well and should come up with a pattern of what we support.

Example:

1. We won't support reading major version differences
2. Minors are backwards, not fwd compat
3. Build versions are backwards and forwards compatible (i.e. no schema change) 

This should also go in a readme if we're going to try and do this (can be added when we actually read this back in however)

Reply:
I'm additionally wondering if the comment-separated sections are going to be painful to parse downstream. Maybe it makes more sense to have a singular root struct which the N components exist in and write a single object? Add the header, no problem, but having a single struct *should* make deserialization a lot more straightforward I would think. In this case, we'll need to identify the section from the comments, and run a different parser each time right?

Reply:
Single object idea SGTM. I guess we'd need an order-preserving serialization/deserialization in that case to make sense out of the generated file.

Reply:
@slawlor  @eozturk1 I've thought about using a single struct to simplify parsing, but the main drawback is that we would have to hold the entire struct in memory until all data is populated and we are ready to write. `serde_yaml` (or `serde_json` for that matter) does not support writing a partial struct.

If we are capturing n states, this would mean holding n directories' worth of db records in memory at one time. This would limit the directory size of fixtures we can generate with the tool.

With the current approach, all I need to do is break the YAML up by the start marker (`---`) to get independent objects. For each chunk, I'll try to parse them as a certain struct, and move on to the next if it fails and only give up after attempting all structs. If this proves hard for some reason, I can experiment with using [YAML tags to denote the struct types](https://stackoverflow.com/questions/15233335/i-dont-understand-what-a-yaml-tag-is).